### ISSUE TITLE
Workers create two python processes for each GPU

### ISSUE THREAD

**Issue by pentschev** [rapidsai/dask-cuda#104]
It has been reported by @pradghos in https://github.com/rapidsai/dask-cuda/issues/74#issuecomment-509687986 that two python processes exist for each GPU when workers get created. Below follows a verbatim copy of the report:

I was trying out to create Multi-node cluster - So started scheduler in another node and tried to create worker using `dask-cuda-worker <scheduler-ip>:8786` on a `4GPU` node. 

From `nvidia-smi` o/p is like below - 
```

|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000004:04:00.0 Off |                    0 |
| N/A   44C    P0    55W / 300W |    620MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000004:05:00.0 Off |                    0 |
| N/A   46C    P0    57W / 300W |    620MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000035:03:00.0 Off |                    0 |
| N/A   42C    P0    54W / 300W |    620MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2...  On   | 00000035:04:00.0 Off |                    0 |
| N/A   48C    P0    56W / 300W |    620MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     66586      C   /opt/anaconda3/bin/python                    305MiB |
|    0     66617      C   /opt/anaconda3/bin/python                    305MiB |
|    1     66586      C   /opt/anaconda3/bin/python                    305MiB |
|    1     66620      C   /opt/anaconda3/bin/python                    305MiB |
|    2     66586      C   /opt/anaconda3/bin/python                    305MiB |
|    2     66618      C   /opt/anaconda3/bin/python                    305MiB |
|    3     66586      C   /opt/anaconda3/bin/python                    305MiB |
|    3     66619      C   /opt/anaconda3/bin/python                    305MiB |
+-----------------------------------------------------------------------------+



$ ps -ef | grep 66586
1084      66586  46641  8 09:48 pts/39   00:00:02 /opt/anaconda3/bin/python /opt/anaconda3/bin/dask-cuda-worker 172.18.0.25:8786
1084      66591  66586  0 09:48 pts/39   00:00:00 [sh] <defunct>
1084      66610  66586  0 09:48 pts/39   00:00:00 /opt/anaconda3/bin/python -c from multiprocessing.semaphore_tracker import main;main(79)
1084      66615  66586  1 09:48 pts/39   00:00:00 /opt/anaconda3/bin/python -c from multiprocessing.forkserver import main; main(94, 108, ['distributed', 'pkg_resources'], **{'sys_path': ['/opt/anaconda3/bin', '/opt/anaconda3/lib/python36.zip', '/opt/anaconda3/lib/python3.6', '/opt/anaconda3/lib/python3.6/lib-dynload', '/opt/anaconda3/lib/python3.6/site-packages']})
pradghos  66675  54389  0 09:48 pts/44   00:00:00 grep --color=auto 66586


$ ps -ef |  grep 66617
1084      66617  66615  3 09:48 pts/39   00:00:01 /opt/anaconda3/bin/python -c from multiprocessing.forkserver import main; main(94, 108, ['distributed', 'pkg_resources'], **{'sys_path': ['/opt/anaconda3/bin', '/opt/anaconda3/lib/python36.zip', '/opt/anaconda3/lib/python3.6', '/opt/anaconda3/lib/python3.6/lib-dynload', '/opt/anaconda3/lib/python3.6/site-packages']})
pradghos  66681  54389  0 09:48 pts/44   00:00:00 grep --color=auto 66617

```

`66586` is started on each GPU and `66617`, `66620`, `66618` and `66619` are probably the worker started on each GPU. 

Question here is - 1> is that expected to see process - `66586` ( `/opt/anaconda3/bin/python /opt/anaconda3/bin/dask-cuda-worker` ) in every GPU and also process-id `66586` has allocated another `305MB` from each GPU which may not be required. 

@pentschev : Any comment / suggestion on this behavior would help. 

Thanks!

**Comment by pentschev** [comment]
Looking at the processes and the PIDs, it looks like 

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 396.26                 Driver Version: 396.26                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:06:00.0 Off |                    0 |
| N/A   37C    P0    56W / 300W |    827MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  Off  | 00000000:07:00.0 Off |                    0 |
| N/A   37C    P0    58W / 300W |    827MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  Off  | 00000000:0A:00.0 Off |                    0 |
| N/A   39C    P0    58W / 300W |    827MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2...  Off  | 00000000:0B:00.0 Off |                    0 |
| N/A   37C    P0    57W / 300W |    827MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   4  Tesla V100-SXM2...  Off  | 00000000:85:00.0 Off |                    0 |
| N/A   37C    P0    57W / 300W |    827MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   5  Tesla V100-SXM2...  Off  | 00000000:86:00.0 Off |                    0 |
| N/A   37C    P0    60W / 300W |    827MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   6  Tesla V100-SXM2...  Off  | 00000000:89:00.0 Off |                    0 |
| N/A   39C    P0    57W / 300W |    827MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   7  Tesla V100-SXM2...  Off  | 00000000:8A:00.0 Off |                    0 |
| N/A   36C    P0    56W / 300W |    827MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      7518      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    0      7668      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    1      7518      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    1      7670      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    2      7518      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    2      7672      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    3      7518      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    3      7674      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    4      7518      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    4      7676      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    5      7518      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    5      7678      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    6      7518      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    6      7680      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    7      7518      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    7      7682      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
+-----------------------------------------------------------------------------+
```

The output of `nvidia-smi` shows that one process is assigned to all GPUs, and besides that one process,  there's another unique process for each GPU.

What happens is that `dask-cuda-worker` creates a CUDA context (using Numba) on all GPUs when it starts, and then it forks the processing creating a new process for each GPU, and all of them end up creating a new CUDA context.

The behavior with `LocalCUDACluster` is a bit different, GPU 0 ends up with two processes, but all other GPUs have only one. See relevant information for `nvidia-smi` below:

```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     27529      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    0     27867      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    1     27869      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    2     27881      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    3     27871      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    4     27877      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    5     27875      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    6     27873      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
|    7     27879      C   ...nda3/envs/rapids-nightly-0.9/bin/python   408MiB |
+-----------------------------------------------------------------------------+
```

To solve this, we can close the CUDA context after the forking has finished, will open a PR to fix that soon.

**Comment by Yumin-Sun-00** [comment]
I recently have the similar problems. We have 9 gpus here, and I always only choose one to train the task. But when running the training script, all gpus got some python processes, that will occupy each gpu about 12 MB. My pytorch version is `1.4.0`.

I played with `CUDA_VISIBLE_DEVICES=8 python train.py --gpu 8` to find where is the problem. Inside my training script, I use pytorch to manually set cuda device to be 8.

This will gives me an error about `invalid device ordinal`. I think it is because my training script is somehow trying to touch all gpus, and `CUDA_VISIBLE_DEVICES ` makes other gpus not visible anymore.

I don't know why everytime it is going to touch all other gpus.
